plugin/pkg/shoot/ingressaddon/admission.go

```diff
@@ -0,0 +1,248 @@
+			if !oldRecords.HasAll(shoot.Spec.Addons.NginxIngress.IngressDNS.AdditionalRecords...) {
+				return admission.NewForbidden(a, fmt.Errorf("additionalRecords %v must not be changed",
+					shoot.Spec.Addons.NginxIngress.IngressDNS.AdditionalRecords))
+			}
+		}
+
+		// Prevent additionalRecords from being set if NginxIngress was not specified before.
+		if addonDefined && (old.Spec.Addons == nil || old.Spec.Addons.NginxIngress == nil) {
+			if len(shoot.Spec.Addons.NginxIngress.IngressDNS.AdditionalRecords) != 0 {
+				return admission.NewForbidden(a, fmt.Errorf("additionalRecords is expected to be empty but found %v",
+					shoot.Spec.Addons.NginxIngress.IngressDNS.AdditionalRecords))
+			}
+		}
+	}
+
+	// This case is only relevant during migration phase.
+	// TODO: remove in future version.
+	shootInDeletion := shoot.ObjectMeta.DeletionTimestamp != nil
+
+	// Generate internal ingress domain if necessary.
+	if addonDefined && !shootInDeletion && len(shoot.Spec.Addons.NginxIngress.IngressDNS.AdditionalRecords) == 0 {
+		generatedDomain, err := d.generateAdditionalRecord()
+		if err != nil {
+			return err
+		}
+		shoot.Spec.Addons.NginxIngress.IngressDNS.AdditionalRecords = []string{generatedDomain}
+	}
+
+	// Generate default ingress domain if necessary.
+	if addonDefined && !shootInDeletion && len(shoot.Spec.Addons.NginxIngress.IngressDNS.StandardRecords) == 0 &&
+		shoot.Spec.DNS.Domain != nil && len(*shoot.Spec.DNS.Domain) > 0 {
+		ingressDomain := utils.GenerateIngressDomain(*shoot.Spec.DNS.Domain)
+		shoot.Spec.Addons.NginxIngress.IngressDNS.StandardRecords = []string{ingressDomain}
+	}
+
+	return nil
+}
+
+func (d *IngressAddon) generateAdditionalRecord() (string, error) {
+	dnsSecret, err := admissionutils.GetInternalDomainSecret(d.secretLister)
+	if err != nil {
+		return "", apierrors.NewInternalError(err)
+	}
+
+	domain := dnsSecret.Annotations[common.DNSDomain]
+	if len(domain) == 0 {
+		return "", apierrors.NewInternalError(fmt.Errorf("could not determine internal domain since annotation %s is not available", common.DNSDomain))
+	}
+
+	var (
+		domainFound     bool
+		generatedDomain string
+	)
+
+	// Try to find an unused internal domain but for not more than 10 rounds.
+	for i := 0; !domainFound && i < 10; i++ {
+		prefixLimit := 11
+		prefix, err := utils.GenerateRandomStringFromCharset(prefixLimit, "0123456789abcdefghijklmnopqrstuvwxyz")
+		if err != nil {
+			return "", apierrors.NewInternalError(err)
+		}
+
+		generatedDomain = fmt.Sprintf("%s.%s", prefix, domain)
+		shoots, err := d.shootLister.List(labels.Everything())
+		if err != nil {
+			return "", apierrors.NewInternalError(err)
+		}
+
+		var domainInUse bool
+		for _, shoot := range shoots {
+			domainInUse = false
+			if shoot.Spec.Addons != nil && shoot.Spec.Addons.NginxIngress != nil {
+				for _, domain := range shoot.Spec.Addons.NginxIngress.IngressDNS.AdditionalRecords {
+					if domain == generatedDomain {
+						domainInUse = true
```

Comment:
Why not directly returning here with a more meaningful error message?

```golang
		return "", apierrors.NewInternalError(errors.New("could not calculate an additionalRecord for Shoot"))
```

Reply:
What do you mean by "calculate again"?

Reply:
If the generated prefix is already used by another Shoot, we generate a new one and check it again (max 10 times).